### DISCUSSION TITLE
Official way to find exact WSL version info from inside the VM?

### DISCUSSION THREAD

**Discussion by K900** [microsoft/WSL#8315] [Q&A]
Basically the question. Specifically asking this in the context of https://github.com/nix-community/NixOS-WSL/, where we need to know the WSL version to get the right flags for the interop binfmt handler since the 0.58 change.

**Comment by Biswa96** [comment]
Would not it be better to parse the required file in /proc/sys/fs/binfmt_misc/ directory?

**Reply by K900** [reply]
No, because interop may be disabled in wsl.conf and enabled in NixOS configuration, in which case we still want to register it.

**Reply by Biswa96** [reply]
Then parse wsl.conf also? 🤔 I am not sure if there is any official docs about but thinking of some workarounds.

1. Get the WSL version from the distro launcher and pass it to WSL2 VM.
2. There is /etc/versions.txt which is mounted as /mnt/wsl/versions.txt but it contains WSLg version.
3. Or get the Linux kernel version. It is somewhat different in WSL preview releases.

**Reply by K900** [reply]
Parsing wsl.conf doesn't give you the version, unfortunately. Also, I don't see any versions.txt in /mnt/wsl or /mnt/wslg.

**Reply by nzbr** [reply]
wsl.conf is generated from the NixOS config anyway, so there is no need to parse it. However, that doesn't help in determining whether preserveArgvZero should be set.

When the preview is installed from the MS Store, it should be possible to get the version through Powershell. That would work until it is not a preview anymore. (The Powershell script could probably extended to take the windows build number into consideration once that's the case)

**Reply by K900** [reply]
Unfortunately, calling into powershell requires working interop in the first place.
